src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6458,7 +6458,7 @@ protected override void VisitForEachExpression(BoundForEachStatement node)
                 SetResultType(expression: null, resultTypeWithState);
             }
 
-            if (reportedDiagnostic || node.EnumeratorInfoOpt == null)
+            if (reportedDiagnostic || node.EnumeratorInfoOpt == null || node.Expression is BoundConversion { Operand: { IsSuppressed: true } })
```

Comment:
It took me some time to verify this is correct. Essentially in the non-error case the expression is always set to a `BoundConversion`. In the error case, which isn't relevant here, it can be a `BoundExpression`.

One thing to consider more generally: what if we made `BoundExpression.Expression` have the type `BoundConversion`? In the error case we could add an identity conversion for the collection type.

Don't want to block this PR on the change but it seems like something to consider.

Reply:
It did feel a little odd for the collection expression to always be a BoundConversion in the "valid" case. Filed #38615 for follow-up.